src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+        offset = put_data_to_the_end(ns_info, datadesc, kval->key, buffer->base_ptr, size);
+        if (0 == offset) {
+            /* this is an error */
+            PMIX_RELEASE(buffer);
+            PMIX_ERROR_LOG(PMIX_ERROR);
+            return PMIX_ERROR;
+        }
+        /* if it's the first time when we put data for this rank, then *rinfo == NULL,
+         * and even if segment was extended, and data was put into the next segment,
+         * we don't need to extension slot at the end of previous segment.
+         * If we try, we might overwrite other segments memory,
+         * because previous segment is already full. */
+        if (free_offset != offset && NULL != *rinfo) {
+            /* here we compare previous free offset with the offset where we just put data.
+             * It should be equal in the normal case. It it's not true, then it means that
+             * segment was extended, and we put data to the next segment, so we now need to
+             * put extension slot at the end of previous segment with a "reference" to a new_offset */
+            size_t sz = sizeof(size_t);
+            addr = _get_data_region_by_offset(datadesc, free_offset);
+            strncpy((char *)addr, "EXTENSION_SLOT", PMIX_MAX_KEYLEN+1);
+            memcpy(addr + PMIX_MAX_KEYLEN + 1, &sz, sizeof(size_t));
+            memcpy(addr + PMIX_MAX_KEYLEN + 1 + sizeof(size_t), &offset, sizeof(size_t));
+        }
+        if (NULL == *rinfo) {
+            *rinfo = (rank_meta_info*)malloc(sizeof(rank_meta_info));
+            (*rinfo)->rank = rank;
+            (*rinfo)->offset = offset;
+            (*rinfo)->count = 0;
+        }
+        (*rinfo)->count++;
+    } else if (NULL != *rinfo) {
+        /* there is data blob for this rank */
+        addr = _get_data_region_by_offset(datadesc, (*rinfo)->offset);
+        if (NULL == addr) {
+            PMIX_RELEASE(buffer);
+            PMIX_ERROR_LOG(PMIX_ERROR);
+            return rc;
+        }
+        /* go through previous data region and find key matches.
+         * If one is found, then mark this kval as invalidated.
+         * Then put a new empty offset to the next extension slot,
+         * and add new kval by this offset.
+         * no need to update meta info, it's still the same. */
+        kval_cnt = (*rinfo)->count;
+        int add_to_the_end = 1;
+        while (0 < kval_cnt) {
+            /* data is stored in the following format:
+             * key[PMIX_MAX_KEYLEN+1]
+             * size_t size
+             * byte buffer containing pmix_value, should be loaded to pmix_buffer_t and unpacked.
+             * next kval pair
+             * .....
+             * extension slot which has key = EXTENSION_SLOT and a size_t value for offset to next data address for this process.
+             */
+            if (0 == strncmp((const char *)addr, "EXTENSION_SLOT", PMIX_MAX_KEYLEN+1)) {
+                offset = *(size_t *)(addr + PMIX_MAX_KEYLEN + 1 + sizeof(size_t));
```

Comment:
I suggest to use the `size` field of extention slot to store the `offset` and consider size to be 0.